schema/crdb/dbinit.sql

```diff
@@ -1480,32 +1480,33 @@ CREATE TABLE IF NOT EXISTS omicron.public.ip_pool (
     time_modified TIMESTAMPTZ NOT NULL,
     time_deleted TIMESTAMPTZ,
 
-    /* Identifies if the IP Pool is dedicated to Control Plane services */
-    internal BOOL NOT NULL,
-
     /* The collection's child-resource generation number */
     rcgen INT8 NOT NULL,
 
+    "default" BOOLEAN NOT NULL DEFAULT FALSE,
+
     /*
-     * Fields representating association with a silo or project. silo_id must be
-     * non-null if project_id is non-null. When project_id is non-null, silo_id
-     * will (naturally) be the ID of the project's silo. Both must be null if
-     * internal is true, i.e., internal IP pools must be fleet-level pools.
+     * Fields representating association with a silo or project. silo_id must
+     * be non-null if project_id is non-null. silo_id is also used to mark an IP
+     * pool as "internal" by associating it with the oxide-internal silo.
      */
     silo_id UUID,
     project_id UUID,
 
     -- if silo_id is null, then project_id must be null
     CONSTRAINT project_implies_silo CHECK (
       NOT ((silo_id IS NULL) AND (project_id IS NOT NULL))
-    ),
-
-    -- if internal = true, non-null silo_id and project_id are not allowed 
-    CONSTRAINT internal_pools_have_null_silo_and_project CHECK (
-       NOT (INTERNAL AND ((silo_id IS NOT NULL) OR (project_id IS NOT NULL)))
     )
 );
 
+/*
+ * Ensure there can only be one default pool for the fleet or a given silo or project
+ */
+CREATE UNIQUE INDEX IF NOT EXISTS one_default_pool_per_scope ON omicron.public.ip_pool (
+    silo_id, project_id
+) WHERE
+    "default" = true AND time_deleted IS NULL;
+
```

Comment:
Need to make sure this index treats `null` as a relevant distinct value, so that, e.g., at the `(null, null)` level (the fleet-wide level) there can only be one default pool.

Reply:
Welp, it doesn't work if either or both is null because it doesn't treat null the way it treats normal values. Looks like this might work as a workaround:

```sql
CREATE OR REPLACE FUNCTION f_non_null_id(id uuid)
RETURNS uuid AS $$
BEGIN
  RETURN COALESCE(id, '00000000-0000-0000-0000-000000000000'::uuid);
END;

CREATE UNIQUE INDEX IF NOT EXISTS one_default_pool_per_scope ON omicron.public.ip_pool (
    f_non_null_id(silo_id), f_non_null_id(project_id)
) WHERE
    "default" = true AND time_deleted IS NULL;
```

_except_ CRDB doesn't support [`CREATE FUNCTION`](https://www.cockroachlabs.com/docs/v22.2/create-function) until 22.2.

<img width="579" alt="Screenshot 2023-08-29 at 11 24 49 PM" src="https://github.com/oxidecomputer/omicron/assets/3612203/e9cbf58a-2f68-4f96-b5e1-d8958e3e4f46">